### PR TITLE
apps/examples/st_things/light: Enable notify switch resource observers manually in sample light things

### PR DESCRIPTION
**Pull request by harish-191**
Add example of sample_light_things to enable notify switch resource observers manually using switch(SW702) of artik053 board and change the state of light by press SW702.
After that immediately notify to observer the state of light.

Signed-off-by: harish-191 <harish.191@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [840b1d89598187e666fa343041a78e96efb16c5a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [840b1d89598187e666fa343041a78e96efb16c5a] - Code Rule Check Result:
apps/examples/st_things/light/sample_light_things.c:132: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/st_things/light/sample_light_things.c:132: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [c9cacf076027db7e8ffe9f2e769722fcf4238674] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c9cacf076027db7e8ffe9f2e769722fcf4238674] - Code Rule Check OK.

File: apps/examples/st_things/light/sample_light_things.c [hunk 1]
```diff
@@ -117,8 +126,12 @@ int ess_process(void)
 
 	iotbus_gpio_context_h m_gpio = iotbus_gpio_open(CONFIG_RESET_BUTTON_PIN_NUMBER);
 	iotbus_gpio_register_cb(m_gpio, IOTBUS_GPIO_EDGE_RISING, gpio_callback_event, (void *)m_gpio);
-#endif
 
+	//gpio event
+	iotbus_gpio_context_h e_gpio = iotbus_gpio_open(CONFIG_HANDLE_GPIO_EVENT_NUMBER);
+        iotbus_gpio_register_cb(e_gpio, IOTBUS_GPIO_EDGE_RISING, gpio_callback_event_push_services, (void *)e_gpio);
```
Comment (JeongJunSik) [inline apps/examples/st_things/light/sample_light_things.c]: please check the tab

File: build/configs/artik053/st_things/defconfig [hunk 2]
```diff
@@ -1163,6 +1163,7 @@ CONFIG_EXAMPLES_ST_THINGS_LIGHT=y
 # CONFIG_EXAMPLES_ST_THINGS_MSG_HANDLING is not set
 CONFIG_RESET_BUTTON=y
 CONFIG_RESET_BUTTON_PIN_NUMBER=44
+CONFIG_HANDLE_GPIO_EVENT_NUMBER=42
```
Comment (JeongJunSik) [inline build/configs/artik053/st_things/defconfig]: please add kconfig in sample_light

Comment (Hyunjun85) [reply build/configs/artik053/st_things/defconfig]: Could you please explain more detail? 
You mean CONFIG_HANDLE_GPIO_EVENT_NUMBER should be able to set any PIN Number by user?

**Comment by Hyunjun85** [comment]
@harish-191 Could you please check why CI-tool was fail.

File: build/configs/artik053/st_things/defconfig (hunk 2 above)
Comment (harish-191) [reply build/configs/artik053/st_things/defconfig]: I saw in description of artik053 board. The default value for  switch SW702 is gpio no 42.

File: apps/examples/st_things/light/sample_light_things.c (hunk 1 above)
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_things.c]: fixed.

**Comment by seinfra** [comment]
Target : [2999b2f51417d508a1f8b1b9896bb1ae6cf4a1db] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [2999b2f51417d508a1f8b1b9896bb1ae6cf4a1db] - Code Rule Check (C++) OK.

File: apps/examples/st_things/Kconfig [hunk 3]
```diff
@@ -52,6 +52,13 @@ config RESET_BUTTON_PIN_NUMBER
 		The default is the reset button pin number 44 for the ARTIK 053.
 		If you use another board, you can set the pin number for the board.
 
+config CONFIG_HANDLE_GPIO_EVENT_NUMBER
```
Comment (sunghan-chang) [inline apps/examples/st_things/Kconfig]: This has a wrong name. You should remove CONFIG_.

File: apps/examples/st_things/Kconfig [hunk 4]
```diff
@@ -52,6 +52,13 @@ config RESET_BUTTON_PIN_NUMBER
...
+	int "GPIO Number of handle push service"
+	default 42
+	depends on RESET_BUTTON
```
Comment (sunghan-chang) [inline apps/examples/st_things/Kconfig]: Why does this have a dependancy with RESET_BUTTON?

File: apps/examples/st_things/light/sample_light_things.c [hunk 5]
```diff
@@ -104,7 +113,7 @@ static bool handle_set_request(st_things_set_request_message_s *req_msg, st_thin
 	}
 
 	return false;
-}
+}
```
Comment (sunghan-chang) [inline apps/examples/st_things/light/sample_light_things.c]: Remove a white space.

File: apps/examples/st_things/Kconfig (hunk 3 above)
Comment (sunghan-chang) [reply apps/examples/st_things/Kconfig]: Is this event number? Or GPIO number?

Comment (harish-191) [reply apps/examples/st_things/Kconfig]: this is GPIO number.

File: apps/examples/st_things/Kconfig (hunk 4 above)
Comment (harish-191) [reply apps/examples/st_things/Kconfig]: I think it will be depend on EXAMPLES_ST_THINGS_LIGHT

Comment (harish-191) [reply apps/examples/st_things/Kconfig]: I think, it should depend on EXAMPLES_ST_THINGS_LIGHT

File: apps/examples/st_things/Kconfig (hunk 3 above)
Comment (harish-191) [reply apps/examples/st_things/Kconfig]: fixed.

**Comment by seinfra** [comment]
Target : [d91873c2572bc5a9c172e3d6918f54003e1483df] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d91873c2572bc5a9c172e3d6918f54003e1483df] - Code Rule Check OK.

File: apps/examples/st_things/light/sample_light_user.c [hunk 6]
```diff
@@ -34,6 +34,7 @@ static const char *PROPERTY_VALUE_COLOR_TEMPERATURE = "ct";
 
 static char *power_status[] = { "off", "on" };
 
+static char *power;
```
Comment (Hyunjun85) [inline apps/examples/st_things/light/sample_light_user.c]: Clould you please apply rule of variable naming. 
ex) g_

File: apps/examples/st_things/light/sample_light_user.c [hunk 7]
```diff
@@ -34,6 +34,7 @@ static const char *PROPERTY_VALUE_COLOR_TEMPERATURE = "ct";
 
 static char *power_status[] = { "off", "on" };
```
Comment (Hyunjun85) [inline apps/examples/st_things/light/sample_light_user.c]: Could you please match rule of variable naming also.
If it is global variable, begins 'g_'.
ex) g_power_status

File: apps/examples/st_things/Kconfig [hunk 8]
```diff
@@ -52,6 +52,13 @@ config RESET_BUTTON_PIN_NUMBER
 		The default is the reset button pin number 44 for the ARTIK 053.
 		If you use another board, you can set the pin number for the board.
 
+config EXAMPLE_ST_THINGS_SAMPLE_LIGHT_HANDLE_GPIO_NUM
```
Comment (sunghan-chang) [inline apps/examples/st_things/Kconfig]: Name looks not good. 
1. I don't know what this is by name.  What's the meaning of LIGHT_HANDLE_GPIO?
2. Others have EXAMPLES. Why is only this EXAMPLE?
3. Why do you add SAMPLE? Another is not a sample but this is?

File: apps/examples/st_things/light/sample_light_things.c [hunk 9]
```diff
@@ -102,9 +109,8 @@ static bool handle_set_request(st_things_set_request_message_s *req_msg, st_thin
 	} else if (0 == strncmp(req_msg->resource_uri, g_res_temp, strlen(g_res_temp))) {
 		return handle_set_request_on_ct(req_msg, resp_rep);
 	}
-
 	return false;
-}
+}
```
Comment (sunghan-chang) [inline apps/examples/st_things/light/sample_light_things.c]: Missed my comment.

Comment (harish-191) [reply apps/examples/st_things/light/sample_light_things.c]: fixed.

File: apps/examples/st_things/light/sample_light_user.c (hunk 7 above)
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_user.c]: fixed.

File: apps/examples/st_things/light/sample_light_user.c (hunk 6 above)
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_user.c]: fixed.

**Comment by seinfra** [comment]
Target : [d70f95067d028ba94da744e8d2c6d8a8d2ef2aaa] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d70f95067d028ba94da744e8d2c6d8a8d2ef2aaa] - Code Rule Check (C++) OK.

File: build/configs/artik053/st_things/defconfig [hunk 10]
```diff
@@ -1159,6 +1159,7 @@ CONFIG_EXAMPLES_SLSIWIFI_STACKSIZE=2048
 # CONFIG_EXAMPLES_SMART_TEST is not set
 CONFIG_EXAMPLES_ST_THINGS=y
 CONFIG_EXAMPLES_ST_THINGS_LIGHT=y
+CONFIG_LIGHT_SWITCH_BUTTON_PIN_NUMBER=42
```
Comment (SenthilKumarGS) [inline build/configs/artik053/st_things/defconfig]: Add another config parameter like 'CONFIG_RESET_BUTTON=y'.
Ex: CONFIG_LIGHT_SWITCH_BUTTON=y

File: apps/examples/st_things/light/sample_light_user.c [hunk 11]
```diff
@@ -57,23 +58,34 @@ bool handle_get_request_on_switch(st_things_get_request_message_s *req_msg, st_t
...
-		resp_rep->set_str_value(resp_rep, PROPERTY_VALUE_SWITCH, power_status[g_switch_value]);
+		resp_rep->set_str_value(resp_rep, PROPERTY_VALUE_SWITCH, g_power_status[g_switch_value]);
 		st_things_notify_observers(req_msg->resource_uri);
 	}
 
 	printf("[%s]OUT-handle_set_request_on_switch() called..\n", TAG);
 	return true;
 }
 
+//change value of switch using switch(702) at artik board
```
Comment (SenthilKumarGS) [inline apps/examples/st_things/light/sample_light_user.c]: 1.  Add a space after // and first letter of the comment should be upper case.
2.  Change the function name to make it more meaningful. Ex: change_switch_value.

File: apps/examples/st_things/light/sample_light_user.c [hunk 12]
```diff
@@ -57,23 +58,34 @@ bool handle_get_request_on_switch(st_things_get_request_message_s *req_msg, st_t
...
+void change_val(void)
+{
+	printf("[%s] current power value: %s\n", TAG, g_power); 
+	if (g_switch_value  == 0) {
```
Comment (SenthilKumarGS) [inline apps/examples/st_things/light/sample_light_user.c]: Minor: Remove extra space after 'g_switch_value'.

**Comment by sunghan-chang** [comment]
@harish-191 You modified only PR title for my comment. But important thing is a commit, not PR. This is an opensource project and user uses commit history, not PR. When you run git log, commit is shown.
Please modify the commit title.

File: apps/examples/st_things/light/sample_light_things.c [hunk 13]
```diff
@@ -50,6 +51,13 @@ static void gpio_callback_event(void *user_data)
 	printf("gpio_callback_event!!\n");
 	printf("reset :: %d\n", st_things_reset());
 }
+
+static void gpio_callback_event_push_services(void *user_data)
```
[the remark this answers is not in the document]
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_things.c]: fixed.

File: apps/examples/st_things/light/sample_light_things.c [hunk 14]
```diff
@@ -114,14 +120,14 @@ int ess_process(void)
 		return 0;
 	}
 	iotapi_initialize();
-
 	iotbus_gpio_context_h m_gpio = iotbus_gpio_open(CONFIG_RESET_BUTTON_PIN_NUMBER);
 	iotbus_gpio_register_cb(m_gpio, IOTBUS_GPIO_EDGE_RISING, gpio_callback_event, (void *)m_gpio);
+	//gpio event
+	iotbus_gpio_context_h e_gpio = iotbus_gpio_open(CONFIG_LIGHT_SWITCH_BUTTON_PIN_NUMBER);
```
[the remark this answers is not in the document]
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_things.c]: fixed.

File: apps/examples/st_things/light/sample_light_user.c (hunk 11 above)
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_user.c]: fixed.

File: apps/examples/st_things/light/sample_light_user.c (hunk 12 above)
Comment (harish-191) [reply apps/examples/st_things/light/sample_light_user.c]: fixed.

File: build/configs/artik053/st_things/defconfig (hunk 10 above)
Comment (harish-191) [reply build/configs/artik053/st_things/defconfig]: done.

**Comment by seinfra** [comment]
Target : [c5fa3e8937cf2fb9df5f31505c85c53521dea1d7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c5fa3e8937cf2fb9df5f31505c85c53521dea1d7] - Code Rule Check OK.

File: apps/examples/st_things/light/sample_light_things.c [hunk 15]
```diff
@@ -130,7 +141,7 @@ int ess_process(void)
 	if (st_things_start() == ST_THINGS_ERROR_NONE) {
 		printf("[%s]=====================================================\n", TAG);
 		printf("[%s]                    Stack Started                    \n", TAG);
-		printf("[%s]=====================================================\n", TAG);
+		printf("[%s]==============================================i=======\n", TAG);
```
Comment (SenthilKumarGS) [inline apps/examples/st_things/light/sample_light_things.c]: Kindly remove this 'i'.

**Comment by seinfra** [comment]
Target : [b9ee64f4291276e698606cbb2f8df08d5ffacf15] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [b9ee64f4291276e698606cbb2f8df08d5ffacf15] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1b3767dd7173322cecc6894285abf49ddac9f115] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1b3767dd7173322cecc6894285abf49ddac9f115] - Code Rule Check OK.

File: apps/examples/st_things/light/sample_light_things.c [hunk 16]
```diff
@@ -52,6 +53,15 @@ static void gpio_callback_event(void *user_data)
 }
 #endif
 
+#ifdef CONFIG_LIGHT_SWITCH_BUTTON
+static void gpio_callback_event_trigger_light_switch(void *user_data)
+{
+	printf("gpio call back event for notify switch resource observers manually!!\n");
```
Comment (SenthilKumarGS) [inline apps/examples/st_things/light/sample_light_things.c]: Change the text to: gpio_callback_event to notify switch resource's observers manually

**Comment by seinfra** [comment]
Target : [460eb6828f1a3d2075029d1a4855f470d75eea7f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [460eb6828f1a3d2075029d1a4855f470d75eea7f] - Code Rule Check OK.

**Pull request merged**